src/services/oss/backend.rs

```diff
@@ -0,0 +1,977 @@
+        builder.build()
+    }
+}
+
+impl Backend {
+    fn oss_put_object_request(
+        &self,
+        path: &str,
+        size: Option<u64>,
+        body: AsyncBody,
+    ) -> Result<Request<AsyncBody>> {
+        let p = build_abs_path(&self.root, path);
+
+        let url = format!("{}/{}", self.endpoint, percent_encode_path(&p));
+
+        let mut req = Request::put(&url);
+
+        req = req
+            .header(HOST, &self.host)
+            .header(CONTENT_TYPE, "application/octet-stream");
+
+        if let Some(size) = size {
+            req = req.header(CONTENT_LENGTH, size)
+        }
+
+        let req = req
+            .body(body)
+            .map_err(|e| new_request_build_error(Operation::Write, path, e))?;
+        Ok(req)
+    }
+
+    fn oss_get_object_request(
+        &self,
+        path: &str,
+        offset: Option<u64>,
+        size: Option<u64>,
+    ) -> Result<Request<AsyncBody>> {
+        let p = build_abs_path(&self.root, path);
+
+        let url = format!("{}/{}", self.endpoint, percent_encode_path(&p));
+
+        let mut req = Request::get(&url);
+        req = req
+            .header("Host", &self.host)
+            .header(CONTENT_TYPE, "application/octet-stream");
+
+        if offset.unwrap_or_default() != 0 || size.is_some() {
+            req = req.header(RANGE, BytesRange::new(offset, size).to_string());
+        }
+
+        let req = req
+            .body(AsyncBody::Empty)
+            .map_err(|e| new_request_build_error(Operation::Read, path, e))?;
+
+        Ok(req)
+    }
+
+    fn oss_delete_object_request(&self, path: &str) -> Result<Request<AsyncBody>> {
+        let p = build_abs_path(&self.root, path);
+
+        let url = format!("{}/{}", self.endpoint, percent_encode_path(&p));
+
+        let mut req = Request::delete(&url);
+        req = req.header("Host", &self.host);
+
+        let req = req
+            .body(AsyncBody::Empty)
+            .map_err(|e| new_request_build_error(Operation::Delete, path, e))?;
+
+        Ok(req)
+    }
+
+    fn oss_head_object_request(&self, path: &str) -> Result<Request<AsyncBody>> {
+        let p = build_abs_path(&self.root, path);
+
+        let url = format!("{}/{}", self.endpoint, percent_encode_path(&p));
+
+        let mut req = Request::head(&url);
+        req = req.header(HOST, &self.host);
+
+        let req = req
+            .body(AsyncBody::Empty)
+            .map_err(|e| new_request_build_error(Operation::Stat, path, e))?;
+
+        Ok(req)
+    }
+
+    fn oss_list_object_request(
+        &self,
+        path: &str,
+        token: Option<String>,
+    ) -> Result<Request<AsyncBody>> {
+        let p = build_abs_path(&self.root, path);
+
+        let url = format!("{}/?list-type=2", self.endpoint,);
+
+        let req = Request::get(&url)
+            .header(HOST, &self.host)
+            .header("delimiter", "/")
+            .header("prefix", p)
+            .header("continuation-token", token.unwrap_or_default())
```

Comment:
header? Are you sure?

Reply:
This avoids signing problems on the listing, but it does not really fit RESTful API.

Reply:
Negetive. It's wrong.